chain/lf3/ec.go

```diff
@@ -0,0 +1,204 @@
+package lf3
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-f3"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/power"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/vm"
+)
+
+type ecWrapper struct {
+	ChainStore   *store.ChainStore
+	StateManager *stmgr.StateManager
+	Manifest     f3.Manifest
+}
+
+type tsWrapper struct {
+	inner *types.TipSet
+}
+
+func (ts *tsWrapper) Key() gpbft.TipSetKey {
+	return ts.inner.Key().Bytes()
+}
+
+func (ts *tsWrapper) Beacon() []byte {
+	entries := ts.inner.Blocks()[0].BeaconEntries
+	if len(entries) == 0 {
+		return []byte{}
```

Comment:
Because https://github.com/filecoin-project/go-f3/blob/69e5251b1d1e8ff91c48b5b1c7527ee84e81ebe3/gpbft/message_builder.go#L121-L123? Because if that's the case, there needs to be a comment here.

Reply:
Yes I believe so. Otherwise CONVERGE message validations would fail. For now I'll comment and leave a TODO.